pkg/internal/server/grpc/common_test.go

```diff
@@ -204,62 +234,176 @@ func (th *testHarness) makeFakeSnapshotClient() *fakesnapshot.Clientset {
 	fakeSnapshotClient := fakesnapshot.NewSimpleClientset()
 	fakeSnapshotClient.PrependReactor("get", "volumesnapshots", func(action clientgotesting.Action) (handled bool, ret apiruntime.Object, err error) {
 		ga := action.(clientgotesting.GetAction)
-		vs := volumeSnapshot(ga.GetName(), ga.GetNamespace())
+		vs := th.VolumeSnapshot(ga.GetName(), ga.GetNamespace())
 		return true, vs, nil
 	})
 	fakeSnapshotClient.PrependReactor("get", "volumesnapshotcontents", func(action clientgotesting.Action) (handled bool, ret apiruntime.Object, err error) {
 		ga := action.(clientgotesting.GetAction)
-		vsc := volumeSnapshotContent(ga.GetName(), th.DriverName)
+		vsc := th.VolumeSnapshotContent(ga.GetName(), th.DriverName)
 		return true, vsc, nil
 	})
+	fakeSnapshotClient.PrependReactor("get", "volumesnapshotclasses", func(action clientgotesting.Action) (handled bool, ret apiruntime.Object, err error) {
+		ga := action.(clientgotesting.GetAction)
+		if ga.GetName() != th.VolumeSnapshotClassName {
+			return false, nil, nil
+		}
+		vs := th.VSCNoAnn() // Not the default but parameters set
+		vs.Name = ga.GetName()
+		return true, vs, nil
+	})
 	return fakeSnapshotClient
 }
 
-func volumeSnapshot(name, namespace string) *snapshotv1.VolumeSnapshot {
+// FakeCountReactors saves the length of the fake reaction chains.
+// Implicitly called by WithFakeClientAPIs but can be explicitly
+// called if additional "permanent" reactors are added.
+func (th *testHarness) FakeCountReactors() {
+	th.lenFakeKubeClientReactionChain = len(th.FakeKubeClient.ReactionChain)
+	th.lenFakeSnapshotClientReactionChain = len(th.FakeSnapshotClient.ReactionChain)
+	th.lenFakeCBTClientReactionChain = len(th.FakeCBTClient.ReactionChain)
+}
+
+// FakePopPrependedReactors pops reactors from the front of the fake reaction chains.
+// Do not use if AddReactor was used after the call to CountFakeReactors.
+func (th *testHarness) FakePopPrependedReactors() {
+	var numToPop int
+
+	numToPop = len(th.FakeKubeClient.ReactionChain) - th.lenFakeKubeClientReactionChain
+	if numToPop > 0 {
+		th.FakeKubeClient.ReactionChain = th.FakeKubeClient.ReactionChain[numToPop:]
+	}
+
+	numToPop = len(th.FakeSnapshotClient.ReactionChain) - th.lenFakeSnapshotClientReactionChain
+	if numToPop > 0 {
+		th.FakeSnapshotClient.ReactionChain = th.FakeSnapshotClient.ReactionChain[numToPop:]
+	}
+
+	numToPop = len(th.FakeCBTClient.ReactionChain) - th.lenFakeCBTClientReactionChain
+	if numToPop > 0 {
+		th.FakeCBTClient.ReactionChain = th.FakeCBTClient.ReactionChain[numToPop:]
+	}
+}
```

Comment:
These are used to remove any Prepended reactors in test cases, so that setting a reactor in one test case does not impact a later test case that utilizes the same fake reaction chain.

Reply:
How does it work? Does it always remove the oldest reactor (FIFO)? This won't be useful if we want to pop specific reactor which might cause conflict.

Reply:
Its handled like a stack (LIFO) - the `PrependReactor` call
```
func (c *Fake) PrependReactor(verb, resource string, reaction ReactionFunc) {
	c.ReactionChain = append([]Reactor{&SimpleReactor{verb, resource, reaction}}, c.ReactionChain...)
}
```
pushes the new reactor to the front of the chain. I'm simply left-shifting these off with `ReactorChain[numToPop:]`.